### PR TITLE
Update program algorithm

### PR DESCRIPTION
**Pull request by StaGlo**
Major algorithm changes.
Improved events handling and performance.
Reduce communication complexity.
Introduce necessary variables.

### CODE REVIEW

**Comment by StaGlo** [comment]
@TypicalAM, we'd appreciate your thoughts on the proposed changes. 

Your feedback will be thoroughly considered.

File: README.md [hunk 1]
```diff
@@ -1,38 +1,105 @@
...
+W pewnym lesie żyje sobie spokojnie Z zajączków i N niedźwiedzi. Co pewien czas zajączki urządzają imprezy. W tym celu uzgadniają między sobą, na której z P nierozróżnialnych polan odbędzie się impreza. Na polanie zmieści się S zajączków, przy czym jeden niedźwiedź zajmuje tyle miejsca, co 4 zajączki.
+
+Każdy zajączek przynosi na imprezę alkohol (to nie zasób, po prostu przed imprezą zajączek deklaruje, co przynosi)Jeżeli na imprezie pojawi się niedźwiedź, zajączki muszą zezwolić mu na picie na sępa, co oznacza, że muszą przynieść dodatkowy alkohol także dla misiów. Impreza z samymi niedźwiedziami nie może się odbyć, bo nie ma kto przynieść alkoholu.
+
+Założenia: 
+```math
+P * (S - 1) < Z
+```
+ 
+Z rozszerzonej zasady szufladkowej Dirichleta, pewne jest, że jakaś impreza dojdzie do skutku, wtedy i tylk wtedy, gdy liczba zajączków jest większa od iloczynu liczby polan i pojemności polan pomniejszonej o 1.
```
Comment (TypicalAM) [inline README.md]: Should be "i tylko wtedy"

**Review by TypicalAM — CHANGES_REQUESTED** [review]
LGTM for the most part!

File: README.md [hunk 2]
```diff
@@ -1,38 +1,105 @@
...
+- Pozostaje w stanie **WAIT** tak długo aż nie uzyska wystarczającej liczby pozwoleń, a następnie przechodzi do stanu **GLADE** wysyłając informację `ENTER`
+- Reakcja na wiadomości:
+    - `REQ`: jeśli pryjorytet żądania jest mniejszy (zegar Lamporta ma większą wartość albo równą, ale większy numer PESEL) zwierzę dokłada żądanie do listy *Que* i nie odpowiada. W przeciwnym wypadku odsyła `ACK`
+    - `ACK`: zwiększa zmienną *Cnt* o jeden w przypadku opowiedzi od zajączka albo o 4 gdy jest ona od niedźwiedzia i nic nie odpowiada
+    - każda inna wiadomość jest nie możliwa, więc ignorowana.
+
+### Przynoszenie alkoholu
+---
+- Zwierzę czeka w stanie **GLADE** do momenut zapełnienia się polany.
+- Gdy wykryje że lista zwierząt biorących udział w imprezie jest pełna wysyła do wszystkich uczestników wiadomość `ALCO`, zerują licznik `alco_count` i przechodzi do stanu **MOREALCO**
```
Comment (rafixxx4k) [inline README.md]: Since alco_count is a variable should be in *italics*

File: README.md [hunk 3]
```diff
@@ -1,38 +1,105 @@
...
+
+### Przynoszenie alkoholu
+---
+- Zwierzę czeka w stanie **GLADE** do momenut zapełnienia się polany.
+- Gdy wykryje że lista zwierząt biorących udział w imprezie jest pełna wysyła do wszystkich uczestników wiadomość `ALCO`, zerują licznik `alco_count` i przechodzi do stanu **MOREALCO** 
+- Reakcja na wiadomości:
+    - `REQ`: dodaje do kolejki *Que*
+    - `ACK`: ignoruje
+    - `ALCO`: jeli pytanie odnosi się do danej polany odpowiada `OK` i przechodzi do stanu **SELFALCO**
+    - `ENTER`: zwiększenie licznika `enter_count`
```
Comment (rafixxx4k) [inline README.md]: Same as above

**Review by rafixxx4k — CHANGES_REQUESTED** [review]

**Comment by StaGlo** [comment]
I have fixed all the issues mentioned above.
Please, review the code and approve if you thinks it's fine.

**Review by rafixxx4k — APPROVED** [review]

**Pull request merged**